File: solutions/1895-minimum-number-of-operations-to-move-all-balls-to-each-box/solution.cpp
class Solution {
public:
    vector<int> minOperations(string boxes) {
        vector<int> leftpass(boxes.size(),0);
        vector<int> rightpass(boxes.size(),0);
        vector<int> ans(boxes.size(),0);
        int gap = 0;
        for(int i=0;i<boxes.size();i++){
            if(i==0){
                rightpass[i] = gap;
            }
            else{
                rightpass[i]= rightpass[i-1]+gap;
            }
            if(boxes[i]=='1'){
                gap++;
            }
         
        }
   

        gap = 0;
        for(int i=boxes.size()-1;i>=0;i--){
            if(i==boxes.size()-1){
                leftpass[i] = gap;
            }
            else{
                leftpass[i]= leftpass[i+1]+gap;
            }
            if(boxes[i]=='1'){
                gap++;
            }
        }
        

        for(int i=0;i<boxes.size();i++){
            ans[i] = leftpass[i]+rightpass[i];
        }
        return ans;
    }
};

File: solutions/1895-minimum-number-of-operations-to-move-all-balls-to-each-box/solution.py
class Solution:
    def minOperations(self, boxes: str) -> List[int]:
        ans = [0]*len(boxes)
        prefixSum = [0]*len(boxes)
        suffixSum = [0]*len(boxes)
        distance = 0
        for i in range(0,len(boxes)):
            if i==0:
                prefixSum[i]=distance
            else:
                prefixSum[i] = distance+prefixSum[i-1]
            if(boxes[i]=='1'):
                distance+=1

        distance = 0

        for i in range(len(boxes)-1,-1,-1):
            if i==len(boxes)-1:
                suffixSum[i]=distance
            else:
                suffixSum[i] = distance+suffixSum[i+1]
            if(boxes[i]=='1'):
                distance+=1

        for i in range(0,len(boxes)):
            ans[i] = prefixSum[i]+suffixSum[i]
        return ans
        
        
